Create Item repository and controller

lhenz15 opened the pull request.

Commit:
Creating Item repository and controller

Implementing Item repossitory and controller

Adding constructor to ItemController

Generating id on ItemRepository

reynoldmorel commented on src/main/java/repositories/ItemRepository.java:
```diff
@@ -0,0 +1,52 @@
+
+public class ItemRepository implements IGenericRepository<Item> {
+
+    Map<Integer, Item> items;
```
Could make this private?

reynoldmorel replied:
Also I think you can initialize it directly

reynoldmorel commented on src/main/java/repositories/ItemRepository.java:
```diff
@@ -0,0 +1,52 @@
+
+    @Override
+    public void add(Item item) {
+        items.put(item.getId(), item);
```
You should generate the id within the repository

reynoldmorel commented on src/main/java/repositories/ItemRepository.java:
```diff
@@ -0,0 +1,52 @@
+    @Override
+    public Item update(Item item) {
+        if (item != null && items.containsKey(item.getId())) {
+            return items.replace(item.getId(), item);
```
You should throw some sort of item not found exception if item is not found. Also you should return the updated item

reynoldmorel commented on src/main/java/repositories/ItemRepository.java:
```diff
@@ -0,0 +1,52 @@
+    @Override
+    public Item delete(Integer id) {
+        if (id != null && items.containsKey(id)) {
+            return items.remove(id);
```
Same here if item is not found just throw an exception

reynoldmorel replied:
You don't need this check

reynoldmorel commented on src/main/java/repositories/ItemRepository.java:
```diff
@@ -0,0 +1,52 @@
+    @Override
+    public Item find(Integer id) {
+        if (id != null && items.containsKey(id)) {
+            return items.get(id);
```
Use getOrDefsult()

lhenz15 replied:
What should I return by default? an empty Item?

reynoldmorel commented on src/main/java/repositories/ItemRepository.java:
```diff
@@ -0,0 +1,52 @@
+
+    @Override
+    public Item find(Integer id) {
+        if (id != null && items.containsKey(id)) {
```
You don't need this check

reynoldmorel commented on src/main/java/repositories/ItemRepository.java:
```diff
@@ -0,0 +1,52 @@
+
+    @Override
+    public Iterable<Item> findAll() {
+        return items.values();
```
Return a list from here. Just do stream() and collect()

Commit:
Creating EntityNotFoundException

Creating NotFoundException and throwing when not found an Item on the update or delete method

Return a List directly from the findAll implementation on ItemRepository

Creating Item constructors

Implementing NotFoundException in the find method on ItemRepository

Injecting ItemRepository on the constructor

Chaging update and delete methods to an GenericType for id

Changing the return of update, delete and find method to an Optional.ofNullable implementation

Renaming NotFoundException to EntityNotFoundException

Changin generic T to K

reynoldmorel commented on src/main/java/controllers/ItemController.java:
```diff
@@ -0,0 +1,35 @@
+
+    private ItemRepository itemRepository;
+
+    public ItemController() {
```
Inject `itemRepository` throught the constructor.

reynoldmorel commented on src/main/java/repositories/IGenericRepository.java:
```diff
@@ -0,0 +1,13 @@
+
+    T update(T t);
+
+    T delete(Integer id);
```
What if I want the key to be `String`?

reynoldmorel commented on src/main/java/repositories/ItemRepository.java:
```diff
@@ -0,0 +1,54 @@
+
+public class ItemRepository implements IGenericRepository<Item> {
+
+    private Map<Integer, Item> items = new HashMap<>();
```
Shouldn't this be final?

reynoldmorel commented on src/main/java/repositories/ItemRepository.java:
```diff
@@ -0,0 +1,54 @@
+
+    @Override
+    public void add(Item item) {
+        item.setId(itemId);
```
I think you can do `item++` directly here

reynoldmorel commented on src/main/java/repositories/ItemRepository.java:
```diff
@@ -0,0 +1,54 @@
+        if(items.replace(item.getId(), item) != null)
+            return item;
+
+        throw new NotFoundException("Item not found.");
```
Let's rename this to `EntityNotFoundException`

reynoldmorel commented on src/main/java/repositories/ItemRepository.java:
```diff
@@ -0,0 +1,54 @@
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class ItemRepository implements IGenericRepository<Item> {
```
nit:
```
return Optional.ofNullable(items.get(id)).orElseThrow(EntityNotFoundException::new)
```

reynoldmorel commented on src/main/java/utils/NotFoundException.java:
```diff
@@ -0,0 +1,7 @@
+
+public class NotFoundException extends RuntimeException {
+    public NotFoundException(String message) {
+        super(message);
```
Use default message.

reynoldmorel reviewed (approved):
Nit: could you change generic T to K since looks like K represents better the key data type
